File: allcode/1300-1399/1353maxEvents.py
# 给你一个数组 events，其中 events[i] = [startDayi, endDayi] ，表示会议 i 开始于 startDayi ，结束于 endDayi 。
#
# 你可以在满足 startDayi <= d <= endDayi 中的任意一天 d 参加会议 i 。注意，一天只能参加一个会议。
#
# 请你返回你可以参加的 最大 会议数目。
#
#
#
# 示例 1：
#
#
#
# 输入：events = [[1,2],[2,3],[3,4]]
# 输出：3
# 解释：你可以参加所有的三个会议。
# 安排会议的一种方案如上图。
# 第 1 天参加第一个会议。
# 第 2 天参加第二个会议。
# 第 3 天参加第三个会议。
# 示例 2：
#
# 输入：events= [[1,2],[2,3],[3,4],[1,2]]
# 输出：4
#
#
# 提示：
#
# 1 <= events.length <= 105
# events[i].length == 2
# 1 <= startDayi <= endDayi <= 105

from leetcode.allcode.competition.mypackage import *

class Solution:
    def maxEvents(self, events: List[List[int]]) -> int:
        events.sort()
        n = len(events)
        mx = max(x for _, x in events)  # 最后一天
        cur = 0
        ans = 0
        hp = []
        for i in range(mx + 1):
            while cur < n and events[cur][0] <= i:
                heappush(hp, events[cur][::-1])  # [end, start] 放入hp，按end从小到大
                cur += 1
            while hp and hp[0][0] < i:
                heappop(hp)
            if hp:
                heappop(hp)
                ans += 1

        return ans


so = Solution()
print(so.maxEvents([[1,2],[1,2],[1,6],[1,2],[1,2]]))  # 3
print(so.maxEvents([[1,2],[1,2],[3,3],[1,5],[1,5]]))  # 5
print(so.maxEvents([[1,2],[2,3],[3,4]]))
print(so.maxEvents([[1,2],[2,3],[3,4],[1,2]]))

File: c++/src/1000-1999/1353maxEvents.cpp
#include "lc_pub.h"


class Solution {
    
    public:
    int maxEvents(vector<vector<int>>& events) {
        int n=events.size();
        int mn=100'001, mx=0;
        for (int i=0;i<n;i++) {
            mn=min(mn,events[i][0]);
            mx=max(mx,events[i][1]);
        }
        ranges::sort(events, [](const vector<int>& a, const vector<int>& b) {
                        if (a[0]==b[0]) return a[1] < b[1];
                        return a[0] < b[0];
                    });
        priority_queue<vector<int>, vector<vector<int>>,  decltype([](const vector<int>& a, const vector<int>& b) {
                        if (a[1]==b[1]) return a[0] > b[0];
                        return a[1] > b[1];
                    })> hp;

        int next=mn,ans=0;
        int j=0;
        for (int i=mn;i<=mx;i++) {
            while (j<n &&events[j][0]<=i) {
                hp.push(events[j]);
                j++;
            }
            while (hp.size()&&hp.top()[1]<i) hp.pop();
            if (hp.size()) {
                hp.pop();
                ans++;
            }
        }
        return ans;
    }
};
struct Compare {
    bool operator()(int a, int b) const {
        return a > b; // 返回 true 表示 b 的优先级更高
    }
};
int main()
{
    std::cout<<"test let us start! %s" << __cplusplus <<std::endl;
    auto arr=parseGrid("[[1,10],[1,2],[2,3],[2,2],[2,5]]");
    Solution so;
    cout<<so.maxEvents(arr)<<endl;

    std::priority_queue<int, std::vector<int>,less<int>> minHeap;

    // 插入元素
    minHeap.push(5);
    minHeap.push(2);
    minHeap.push(9);
    minHeap.push(1);
    minHeap.push(5);

    // 输出元素
    std::cout << "小顶堆的元素（从最小到最大）：" << std::endl;
    while (!minHeap.empty()) {
        std::cout << minHeap.top() << " "; // 访问优先级最高的元素（最小值）
        minHeap.pop(); // 移除优先级最高的元素
    }
    std::cout << std::endl;
    return 0;
}
